### PR TITLE
Update IOC

### PR DESCRIPTION
**Pull request by jarry-xiao**

### CODE REVIEW

**Review by eugene-chen — APPROVED** [review]
add a test where ioc succeeds w/ these set and one where it fails

File: src/program/processor/new_order.rs [hunk 1]
```diff
@@ -371,6 +371,10 @@ fn process_new_order<'a, 'info>(
                     )?;
                 }
             }
+        } else {
+            // Should never execute as the account loading logic should fail
```
Comment (eugene-chen) [inline src/program/processor/new_order.rs:375]: nit: Should never be reached

File: src/program/processor/new_order.rs [hunk 2]
```diff
@@ -371,6 +371,10 @@ fn process_new_order<'a, 'info>(
...
+            phoenix_log!("WARNING: Vault context was not provided");
```
Comment (eugene-chen) [inline src/program/processor/new_order.rs:376]: Do we have a canonical log format for things that we expect to never get hit? Would be nice so it's easy to set up a log filter and alert.

File: src/state/markets/fifo.rs [hunk 3]
```diff
@@ -822,7 +822,8 @@ impl<
 
         if order_packet.is_expired(current_slot, current_unix_timestamp) {
             phoenix_log!("Order parameters include a last_valid_slot or last_valid_unix_timestamp_in_seconds in the past, skipping matching and posting");
-            return None;
+            // Fail softly if the order is expired
```
Comment (eugene-chen) [inline src/state/markets/fifo.rs:825]: prefer "Don't fail loudly" or something like "Don't fail the entire tx, but this order doesn't make it to the book"

File: src/state/markets/test_market.rs [hunk 4]
```diff
@@ -1402,6 +1402,60 @@ fn test_fok_and_ioc_limit_5() {
...
+                    None,
+                    Some(2),
+                ),
+                &mut record_event_fn,
+                &mut mock_clock_fn,
+            )
+            .unwrap()
+            .1,
+        MatchingEngineResponse::default()
+    );
```
Comment (eugene-chen) [inline src/state/markets/test_market.rs:1457]: Add one more test that includes a successful IOC with TIF params set in the future

**Review by eugene-chen — APPROVED** [review]
Small changes, otherwise LGTM

File: src/state/markets/test_market.rs (hunk 4 above)
Comment (eugene-chen) [reply src/state/markets/test_market.rs:1459]: i added this and verified tests still pass

File: src/state/markets/fifo.rs [hunk 5]
```diff
@@ -1538,10 +1539,15 @@ impl<
                     if order.trader_index != trader_index as u64 {
                         return None;
                     }
-                    (
-                        base_lots_to_remove == order.num_base_lots,
-                        base_lots_to_remove,
-                    )
+                    // If the order is tagged as expired, we remove it from the book regardless of the size.
+                    if order_is_expired {
```
[the remark this answers is not in the document]
Comment (jarry-xiao) [reply src/state/markets/fifo.rs:1543]: Though given that it's a private function where all public interfaces pass this param as false, maybe this is acceptable

File: src/program/processor/new_order.rs (hunk 2 above)
Comment (eugene-chen) [reply src/program/processor/new_order.rs:376]: there's only like 70 `phoenix_log!` in the entire repo, we could audit for consistency

happy to punt

File: src/state/markets/fifo.rs [hunk 6]
```diff
@@ -822,7 +822,8 @@ impl<
 
         if order_packet.is_expired(current_slot, current_unix_timestamp) {
             phoenix_log!("Order parameters include a last_valid_slot or last_valid_unix_timestamp_in_seconds in the past, skipping matching and posting");
-            return None;
+            // Do not fail the transaction if the order is expired, but do not modify the order book
```
Comment (eugene-chen) [inline src/state/markets/fifo.rs:825]: "...but do not place or match the order`

**Review by eugene-chen — APPROVED** [review]

File: src/state/markets/fifo.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (eugene-chen) [reply src/state/markets/fifo.rs:1543]: discussed offline, fine for now

**Pull request merged**